### DISCUSSION TITLE
Custom Adapter question about getUser

### DISCUSSION THREAD

**Discussion by dpeek** [nextauthjs/next-auth#1970] [Help]
Hi there, thank you all for the awesome library that is making my life better :)

I'm building a custom adaptor for a graph database (dgraph) – though it should work with any GraphQL endpoint with the right schema.

My question is about the `getUser` callback – currently this seems to be called internally to get user info from either a session or an account. Unfortunately it makes an assumption about the shape of the session data (namely that it has a userId field). The problem is that I wanted to model this in GraphQL as an "edge" from session, so that the actual field would be `session.user.id`.

Do you feel this field (userId) needs part of the required contract for next-auth? My first thought was that `getUser` could be modified to accept a session instead on a user id, which means we can look up however we want. Or perhaps `getUserBySession(session)` and `getUserByAccount(account)` would be cleaner?

The bonus with GraphQL is that I can simply return all the fields I need on the session and return `session.user` :)

Thank you for your time.

**Comment by DopamineDriven** [comment]
You could try module augmentation in a `types/next-auth.d.ts` file

[OfficialExample](https://next-auth.js.org/getting-started/typescript#module-augmentation)
```ts
import NextAuth from "next-auth"

declare module "next-auth" {
  /**
   * Returned by `useSession`, `getSession` and received as a prop on the `Provider` React Context
   */
  interface Session {
    user: {
      /** The user's postal address. */
      address: string
    }
  }
}
```

What I have locally 
```ts
import NextAuth, { User, Account, Profile } from 'next-auth';
import {
	WordpressUser,
} from '@/graphql/generated/graphql';
declare module 'next-auth' {
	interface Session
		extends Account,
			Profile,
			WordpressUser {}
}
```

**Reply by dpeek** [reply]
It's not so much the type of the object but that nextauth is expecting the field `userId` on the session – I'm wondering if I can change the shape of the session so that the library doesn't rely on that field being there.

**Comment by balazsorban44** [comment]
So roughly, your `getSession` method could look something like this then:

```js
async getSession(sessionToken) {
  const session = await dgraph.query(SessionQuery, {sessionToken})
  //...
  return {
    userId: session.user.id,
    ...session
  }
}
```

I can see how GraphQL specifically could return the whole `user` here, but we are trying to keep these adapters agnostic, and the whole point of the custom adapters is that they adapt to our core, not that we tailor our core to specific adapters.

This is a problem already with the OAuth Providers which can interpret the OAuth spec in so many ways, so we still have logic in our source code that targets specific providers. I would like to get rid of all that through https://github.com/nextauthjs/next-auth/issues/1846.

So at this point, I don't think I would like to add changes that would make GraphQL implementations easier but would have no gain for anything else really, I hope you understand :slightly_smiling_face:
